Initial project deploy dialog

kisantia opened the pull request:
This adds the first iteration for the project deploy dialog UI to choose the connection. The target connection can be chosen using the connection dialog or one of the project's data sources from a dropdown.

Connection with name:
<img width="1271" alt="Screen Shot 2020-05-11 at 11 17 34 AM" src="https://user-images.githubusercontent.com/31145923/81596551-22a84380-9379-11ea-9706-405041d8e867.png">

Connection without name:
<img width="1269" alt="Screen Shot 2020-05-11 at 11 17 03 AM" src="https://user-images.githubusercontent.com/31145923/81596588-305dc900-9379-11ea-9a49-7be9aa002393.png">

Data source:
<img width="1272" alt="Screen Shot 2020-05-11 at 11 16 39 AM" src="https://user-images.githubusercontent.com/31145923/81596605-36ec4080-9379-11ea-9122-0d2820e18677.png">

aaomidi commented on extensions/sql-database-projects/src/controllers/projectController.ts:
```diff
@@ -120,7 +121,8 @@ export class ProjectsController {
 		const project = this.getProjectContextFromTreeNode(treeNode);
-		await this.apiWrapper.showErrorMessage(`Deploy not yet implemented: ${project.projectFilePath}`); // TODO
+		const deployDatabaseDialog = new DeployDatabaseDialog(project);
+		deployDatabaseDialog.openDialog();
```
this is async, either return it or await it.

aaomidi commented on extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:
```diff
@@ -0,0 +1,291 @@
+		this.dialog.content = [this.deployTab];
+	}
+
+	private initializeDeployTab(): void {
```
This method returns instantly, but isn't actually "complete" when the return is done.

You might want to wrap it around a promise.

aaomidi commented on extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:
```diff
@@ -0,0 +1,291 @@
+
+	// only enable Generate Script and Ok buttons if all fields are filled
+	private tryEnableGenerateScriptAndOkButtons(): void {
+		if (this.targetConnectionTextBox!.value && this.targetDatabaseTextBox!.value && this.deployScriptNameTextBox!.value
```
Why the non-null assertion here?

kisantia replied:
It's to get rid of the "Object is possibly undefined" error because the types of these guys is also undefined because they don't get assigned in the constructor

aaomidi replied:
That's fair, as long as you're absolutely certain they won't be undefined by the time you're in this method :)

aaomidi commented on extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:
```diff
@@ -0,0 +1,291 @@
+
+	private initializeDialog(): void {
+		this.initializeDeployTab();
+		this.dialog.content = [this.deployTab];
```
So technically since initializeDeployTab isn't being awaited, this might have inconsistent behavior.

kisantia replied:
Is it bad if the dialog content isn't registered yet? Agent also initializes and adds its tabs like this.

aaomidi replied:
It's definitely creates uncertain behavior. We've made these types of mistakes all over the code, and that would mean Agent probably needs to be fixed too.

kisantia replied:
Ah ok. Is there an example of somewhere that does this correctly that I can look at?

aaomidi replied:
I'm not sure, but I can make a PR onto your branch that addresses some of these comments.

kisantia replied:
sure that would be really helpful! :)

Benjin commented on extensions/sql-database-projects/src/test/deployDatabaseDialog.test.ts:
```diff
@@ -0,0 +1,62 @@
+		const project = new Project(projFilePath);
+
+		const deployDatabaseDialog = new DeployDatabaseDialog(project);
+		should.equal(deployDatabaseDialog.getDefaultDatabaseName(), projFolder);
```
default database name (and script name below) should probably be the project name rather than the folder name.  I've been meaning to add that as a property in Project...

kisantia replied:
added it as a property in Project

Benjin replied:
Thanks!

kisantia commented on extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:
```diff
@@ -0,0 +1,291 @@
+		azdata.window.closeDialog(this.dialog);
+	}
+
+	private async cancel(): Promise<void> {
```
I think it makes sense to leave this as cancel since this doesn't close the dialog since the dialog already handles closing the dialog when cancelButton is clicked. This also doesn't do anything, so it can also just be removed for now.

kisantia commented on extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts:
```diff
@@ -0,0 +1,291 @@
+		}).component();
+
+		editConnectionButton.onDidClick(async () => {
+			this.connection = <azdata.connection.ConnectionProfile><any>await azdata.connection.openConnectionDialog();
```
ah yeah I forgot about the apiwrapper. Switched this to use apiwrapper.

kisantia commented:
I removed the input box for the generated script file path since the current generate script API opens a new document tab with the script rather than saving the script to a file. Updated the description screenshots with this change.
